chirp/drivers/retevis_ha1g.py

```diff
@@ -0,0 +1,2111 @@
+            status.cur += len(new_data_bytes)
+            self.status_fn(status)
+            item_bytes += new_data_bytes[20:-3]
+            page_index += 1
+            success = True
+            break
+        if not success:
+            raise errors.RadioError("radio reported failure")
+    return item_bytes
+
+
+def write_item_current_page_bytes(self, serial, item_Bytes: bytes,
+                                  item: int, status):
+    if not item_Bytes:
+        raise ValueError("item_bytes cannot be empty")
+    max_retries = 5
+    retry_delay = 0.05
+    page_count = math.ceil(len(item_Bytes) / self.write_page_len)
+    if page_count == 0:
+        return True
+    for i in range(0, page_count):
+        for attempt in range(max_retries):
+            current_page_bytes = get_write_item_send_bytes(
+                item, i, page_count,
+                item_Bytes[
+                    i * self.write_page_len: (i + 1) * self.write_page_len])
+            flag, new_data_bytes = handle_connect_ver(
+                current_page_bytes,
+                serial, self.read_page_len)
+            if (not flag
+               or (len(new_data_bytes) > 14 and new_data_bytes[14] != item)):
+                time.sleep(retry_delay)
+                continue
+            status.cur += len(current_page_bytes)
+            self.status_fn(status)
+            break
+
+
+def get_write_item_bytes(all_bytes: bytearray, item):
+    """
+       Get the bytes for a specific memory region
+       - Uses MEMORY_REGIONS_RANGES to find start and length
+    """
+    if item not in MEMORY_REGIONS_RANGES:
+        LOG.debug(f"Unknown memory Range:{item}")
+        return b""
+    start, length = MEMORY_REGIONS_RANGES[item]
+    return all_bytes[start:start+length]
+
+
+def read_item_handle_connect_event(all_bytes: bytearray,
+                                   item_bytes: bytes, item):
+    """
+    Put the given region bytes into the main memory array
+    - Uses MEMORY_REGIONS_RANGES to find start and length
+    """
+    if item not in MEMORY_REGIONS_RANGES:
+        LOG.debug(f"Unknown memory Range:{item}")
+    start, length = MEMORY_REGIONS_RANGES[item]
+    item_len = min(len(item_bytes), length)
+    all_bytes[start:start + item_len] = item_bytes[:item_len]
+    return all_bytes
+
+
+def exit_programming_mode(self):
+    serial = self.pipe
+    try:
+        serial.write(Get_ReadItemBytes(111, 0, 0))
+    except Exception as e:
+        LOG.debug(
+            f"Radio refused to exit programming mode:{e}")
+        raise errors.RadioError(
+            "Radio refused to exit programming mode")
+
+
+def Get_ReadItemBytes(data_type: int, page_index: int,
+                      data_code: int = 2):
+    data_part = (
+        b"RDTP\x01\x00\x00\x00\x00\x00\x00\x00\x08\x00"
+        + data_type.to_bytes(1, "little")
+        + data_code.to_bytes(1, "little")
+        + b"\x00\x00\x01\x00"
+        + page_index.to_bytes(2, "little"))
```

Comment:
This function appears to format a message for the radio that says "read block of data" right?

Please use `struct.pack()` here and not all this bytestring concatenation, like other drivers do. What are all the magic numbers here? 1? 1? 2?